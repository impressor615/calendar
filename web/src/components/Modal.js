import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Button from 'components/Button';

export const ModalHeader = ({ children, onToggle }) => (
  <div className="modal-header">
    <div>{children}</div>
    <Button onClick={onToggle}>&times;</Button>
  </div>
);

ModalHeader.propTypes = {
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export const ModalBody = ({ children }) => (
  <div className="modal-body">
    {children}
  </div>
);

ModalBody.propTypes = {
  children: PropTypes.node.isRequired,
};

export const ModalFooter = ({ children }) => (
  <div className="modal-footer">
    { children }
  </div>
);

ModalFooter.propTypes = {
  children: PropTypes.node.isRequired,
};

export const ModalDialog = ({ children, ...rest }) => (
  <div className={classnames('modal-dialog', rest.className)}>
    { children }
  </div>
);

ModalDialog.propTypes = {
  children: PropTypes.node.isRequired,
};

const Modal = ({ isOpen, children, backdrop, onToggle, ...rest }) => (
  isOpen ? (
    ReactDOM.createPortal(
      <div
        {...rest}
        className={classnames('modal', rest.className)}
      >
        {children}
        { backdrop ? <div className="modal-backdrop" onClick={onToggle} /> : null }
      </div>,
      document.getElementById('modal-root'),
    )
  ) : null
);

Modal.defaultProps = {
  backdrop: true,
  onToggle: () => {},
};

Modal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onToggle: PropTypes.func,
  backdrop: PropTypes.bool,
};

export default Modal;
