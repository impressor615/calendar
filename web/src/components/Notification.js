import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import Button from 'components/Button';
import Modal, { ModalDialog, ModalBody } from 'components/Modal';


const Notification = ({ type, isOpen, message, onToggle }) => (
  <Modal
    isOpen={isOpen}
    backdrop={false}
    className="notification"
  >
    <ModalDialog
      className={classnames({ [type]: type })}
    >
      <ModalBody>
        <div>{ message }</div>
        <Button onClick={onToggle}>&times;</Button>
      </ModalBody>
    </ModalDialog>
  </Modal>
);

Notification.defaultProps = {
  type: 'danger',
};

Notification.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  message: PropTypes.node.isRequired,
  type: PropTypes.oneOf(['danger', 'success'])
};

export default Notification;
