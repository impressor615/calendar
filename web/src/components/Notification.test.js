import React from 'react';

import Notification from 'components/Notification';
import Modal, { ModalDialog, ModalBody } from 'components/Modal';


describe('<Notification />', () => {
  const modalRoot = document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  const body = document.querySelector('body');
  body.appendChild(modalRoot);

  let wrapper;
  const onToggle = jest.fn();
  const props = {
    type: 'danger',
    isOpen: true,
    message: 'error occurred' ,
    onToggle,
  };

  afterEach(() => {
    wrapper.unmount();
  });

  it('should be rendered properly', () => {
    wrapper = mount(<Notification {...props} />);
    expect(wrapper.find(Modal)).toHaveLength(1);
    expect(wrapper.find(ModalDialog)).toHaveLength(1);
    expect(wrapper.find(ModalBody)).toHaveLength(1);
  });

  it('shoud be matched with the snapshot', () => {
    wrapper = mount(<Notification {...props} />);
    expect(wrapper).toMatchSnapshot();
  })
});

