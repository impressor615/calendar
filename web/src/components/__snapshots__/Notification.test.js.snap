// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification /> shoud be matched with the snapshot 1`] = `
<Notification
  isOpen={true}
  message="error occurred"
  onToggle={[MockFunction]}
  type="danger"
>
  <Modal
    backdrop={false}
    className="notification"
    isOpen={true}
    onToggle={[Function]}
  >
    <Portal
      containerInfo={
        <div
          id="modal-root"
        >
          <div
            class="modal notification"
          >
            <div
              class="modal-dialog danger"
            >
              <div
                class="modal-body"
              >
                <div>
                  error occurred
                </div>
                <button
                  class="btn"
                >
                  ×
                </button>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <div
        className="modal notification"
      >
        <ModalDialog
          className="danger"
        >
          <div
            className="modal-dialog danger"
          >
            <ModalBody>
              <div
                className="modal-body"
              >
                <div>
                  error occurred
                </div>
                <Button
                  active={false}
                  color=""
                  onClick={[MockFunction]}
                >
                  <button
                    className="btn"
                    onClick={[MockFunction]}
                  >
                    ×
                  </button>
                </Button>
              </div>
            </ModalBody>
          </div>
        </ModalDialog>
      </div>
    </Portal>
  </Modal>
</Notification>
`;
